influxdb/src/main/scala/akka/stream/alpakka/influxdb/impl/InfluxDBFlowStage.scala

```diff
@@ -0,0 +1,131 @@
+import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.influxdb.InfluxDB
+import org.influxdb.impl.InfluxDBResultMapperHelper
+
+import scala.collection.immutable
+import org.influxdb.BatchOptions
+import org.influxdb.dto.{BatchPoints, Point}
+
+/**
+ * INTERNAL API
+ */
+private[influxdb] class InfluxDBFlowStage[T, C](
+    clazz: Option[Class[T]],
+    influxDB: InfluxDB
+) extends GraphStage[FlowShape[immutable.Seq[InfluxDBWriteMessage[T, C]], immutable.Seq[InfluxDBWriteResult[T, C]]]] {
+  private val in = Inlet[immutable.Seq[InfluxDBWriteMessage[T, C]]]("in")
+  private val out = Outlet[immutable.Seq[InfluxDBWriteResult[T, C]]]("out")
+
+  override val shape = FlowShape(in, out)
+
+  override protected def initialAttributes: Attributes =
+    super.initialAttributes and Attributes(ActorAttributes.IODispatcher)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    clazz match {
+      case Some(c) => new InfluxDBMapperRecordLogic
+      case None => new InfluxDBRecordLogic
+    }
+
+  sealed abstract class InfluxDBLogic extends GraphStageLogic(shape) with InHandler with OutHandler {
+
+    protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit
+
+    setHandlers(in, out, this)
+
+    override def onPull(): Unit = if (!isClosed(in) && !hasBeenPulled(in)) pull(in)
+
+    override def onPush(): Unit = {
+      val messages = grab(in)
+      if (messages.nonEmpty) {
+
+        influxDB.enableBatch(BatchOptions.DEFAULTS)
+        write(messages)
+        val writtenMessages = messages.map(m => new InfluxDBWriteResult(m, None))
+        influxDB.close()
+        push(out, writtenMessages)
+      }
+
+      tryPull(in)
+    }
+  }
+
+  final class InfluxDBRecordLogic extends InfluxDBLogic {
+
+    override protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit =
+      messages
+        .filter {
+          case InfluxDBWriteMessage(_: Point, _, _, _) => {
+            true
+          }
+          case InfluxDBWriteMessage(_: AnyRef, _, _, _) => {
+            failStage(new RuntimeException(s"unexpected type Point required"))
+            false
+          }
+        }
+        .groupBy(im => (im.databaseName, im.retentionPolicy))
+        .map(wm => toBatchPoints(wm._1._1, wm._1._2, wm._2))
+        .foreach(influxDB.write)
+
+  }
+
+  final class InfluxDBMapperRecordLogic extends InfluxDBLogic {
+
+    private val mapperHelper: InfluxDBResultMapperHelper = new InfluxDBResultMapperHelper
+
+    override protected def write(messages: immutable.Seq[InfluxDBWriteMessage[T, C]]): Unit =
+      messages
+        .groupBy(groupByDbRp)
+        .map(convertToBatchPoints)
+        .foreach(influxDB.write)
+
+    def groupByDbRp(im: InfluxDBWriteMessage[T, C]) =
+      (
+        im.databaseName match {
+          case Some(databaseName) => Some(databaseName)
+          case None => Some(mapperHelper.databaseName(im.point))
```

Comment:
Is there anything we can do to enable caching even in this usage? Or will it cache with the first lookup anyway?

Reply:
@ennru 
This in memory grouping certainly has a cost. The goal is to use the batching capabilities of the driver. The requests are batched per database and retention policy.
To make this more efficient I propose to add the optional fields databaseName and retentionPolicy on the InfluxDbWriteSettings. Those fields will override the database/retention policy information per message. If those optional fields are set then all the measurements/points shall be written (and thus batched) to the same database and retention policy as in the settings. Therefore there won't be the need for the grouping above.
Eventually the user will have the one database/retention policy option, or the various database/retention policy combinations.